From 3a189462c754aa4b68d9deb54115374de9049963 Mon Sep 17 00:00:00 2001
From: Yuvarani V <quic_yuvarani@quicinc.com>
Date: Thu, 19 Sep 2024 09:33:37 +0530
Subject: [PATCH] hostapd: Fix Number of simultaneous links in MLE during CAC

Currently, the maximum number of simultaneous links in MLE is
advertised as `num_links - 1`, where `num_links` is the number
of links added to the MLD AP. However, when the 5 GHz band is
in CAC timeout, this results in an incorrect value being
advertised for the maximum number of simultaneous links in
MLE, as the 5 GHz link is not active.

For example, an AP MLD with 3 links (2 GHz, 5 GHz (in CAC timeout)
and 6 GHz) during bringup has `num_links` set to 3.
Consequently, the maximum number of simultaneous links
in MLE is advertised as 2 according to the current code,
despite the 5 GHz link being in CAC timeout.

To Fix this issue, introduce an API to retrieve the active links
of the MLD AP and use it to set the value for the maximum
number of simultaneous links in MLE.

Patch-dependency: None

Patch-work: [PATCH] hostapd: Fix Number of simultaneous links in MLE during CAC

Signed-off-by: Yuvarani V <quic_yuvarani@quicinc.com>
---
 src/ap/hostapd.c        | 18 ++++++++++++++++++
 src/ap/hostapd.h        |  1 +
 src/ap/ieee802_11_eht.c |  2 +-
 3 files changed, 20 insertions(+), 1 deletion(-)

--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -5534,4 +5534,23 @@ void hostapd_bss_free(struct hostapd_dat
 	return;
 }
 
+
+u8 hostapd_get_active_links(struct hostapd_data *hapd)
+{
+	struct hostapd_data *link_bss;
+	u8 active_links = 0;
+
+	if (!hapd || !hapd->conf->mld_ap)
+		return 0;
+
+	for_each_partner_bss(link_bss, hapd) {
+		if (link_bss == hapd || !link_bss->started)
+			continue;
+
+		active_links++;
+	}
+
+	return active_links;
+}
+
 #endif /* CONFIG_IEEE80211BE */
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -937,6 +937,7 @@ int hostapd_mld_remove_link(struct hosta
 bool hostapd_mld_is_first_bss(struct hostapd_data *hapd);
 struct hostapd_data * hostapd_mld_get_first_bss(struct hostapd_data *hapd);
 void hostapd_bss_free(struct hostapd_data *hapd);
+u8 hostapd_get_active_links(struct hostapd_data *hapd);
 int hostapd_link_remove(struct hostapd_data *hapd, u32 count);
 struct hostapd_data *
 hostapd_interfaces_get_hapd(struct hapd_interfaces *interfaces,
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -681,7 +681,7 @@ u8 * hostapd_eid_eht_basic_ml(struct hos
 	max_simul_links = mld_cap & MLE_MLD_CAP_MAX_SIMULTANEOUS_LINK_MASK;
 	tid_to_link_map = (mld_cap & MLE_MLD_CAP_TID_TO_LINK_MAP_MASK) >>
 			  MLE_MLD_CAP_TID_TO_LINK_MAP_SHIFT;
-	active_links = hapd->mld->num_links - 1;
+	active_links = hostapd_get_active_links(hapd);
 
 	if (active_links > max_simul_links) {
 		wpa_printf(MSG_ERROR, "MLD: Error in max simultaneous links, advertised: %x current: %x",
